data-raw/schemas.R

```diff
@@ -0,0 +1,228 @@
+# will be done:
+#
+# ```
+# source("data-raw/schemas.R")
+# #> ✔ Schemas up-to-date!
+# #> ℹ branch: "main"
+# #> ℹ sha: "0163a89cc38ba3846cd829545f6d65c1e40501a6"
+# #> ℹ timestamp: "2024-12-19T16:26:33Z"
+# ```
+#
+# TO CHANGE THE BRANCH, update the environment variable called
+# `HUBUTILS_DEV_BRANCH`:
+#
+# ```
+# Sys.setenv("HUBUTILS_DEV_BRANCH" = "br-v4.0.1")
+# source("data-raw/schemas.R")
+# #> ✔ removing /path/to/hubUtils/inst/schemas
+# #> ✔ Creating inst/schemas/.
+# #> ℹ Fetching the latest version of the schemas from GitHub
+# #> Cloning into '/var/folders/9p/m996p3_55hjf1hc62552cqfr0000gr/T//Rtmp3Q4dnp/file377d
+# #> 71aaebd4'...
+# #> ✔ Copying v4.0.1, v4.0.0, v3.0.1, v3.0.0, v2.0.1, v2.0.0, v1.0.0, v0.0.1, v0.0.0.9,
+# #>  and NEWS.md to inst/schemas
+# #> [ ... snip ... ]
+# #> ✔ Done
+# #> ✔ Schemas up-to-date!
+# #> ℹ branch: "br-v4.0.1"
+# #> ℹ sha: "43b2c8aceb3a316b7a1929dbe8d8ead2711d4e84"
+# #> ℹ timestamp: "2024-12-19T16:40:16Z"
+# ```
+
+# FUNCTIONS -------------------------------------------------------------------
+script_name <- function() {
+  cmd <- commandArgs()
+  basename(sub("--file=", "", cmd[grepl("--file=", cmd, fixed = TRUE)], fixed = TRUE))
+}
+
+check_status <- function(repo_path) {
+  git_stat <- system2(
+    "git",
+    c("-C", repo_path, "status", "--short", "--porcelain"),
+    stdout = TRUE
+  )
+  paths <- dirname(substring(git_stat, 4, nchar(git_stat)))
+  if (any(startsWith(paths, "inst/schemas"))) {
+    cli::cli_abort(
+      c(
+        "New schemas must be committed before pushing."
+      )
+    )
+  }
+}
+
+not_hook <- function() interactive() || script_name() != "pre-push"
+is_hook <- Negate(not_hook)
+
+check_hook <- function(repo_path) {
+  # if this is running as a git hook, then the first thing to do is to make sure
+  # it is up to date with the source material. If it's not, error until it is
+  # fixed.
+  if (not_hook()) {
+    return()
+  }
+  cli::cli_h2("{.strong pre-push}: checking that the hook is up-to-date")
+  hook <- fs::path(repo_path, ".git/hooks/pre-push")
+  if (fs::file_exists(hook)) {
+    schema_script <- fs::path(repo_path, "data-raw/schemas.R")
+    okay <- tools::md5sum(hook) == tools::md5sum(schema_script)
+    if (isTRUE(okay)) {
+      cli::cli_alert_success("OK")
+    } else {
+      cmd <- r"[usethis::use_git_hook("pre-push", readLines(usethis::proj_path("data-raw/schemas.R")))]" # nolint: object_usage_linter
+      cli::cli_abort(
+        c(
+          "git hook outdated",
+          "i" = r"[Use {.code {cmd}} to update your hook.]"
+        )
+      )
+    }
+  } else {
+    cli::cli_alert("No pre-push hook registered")
+  }
+}
+
+get_branch <- function(update_cfg_path) {
+  if (fs::file_exists(update_cfg_path)) {
+    branch <- jsonlite::read_json(update_cfg_path)$branch
+  } else {
+    branch <- "main"
+  }
+  branch
+}
+
+timestamp <- function() {
+  format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
+}
+
+get_latest_commit <- function(branch) {
```

Comment:
Would it be possible to also always check the main branch and to also flag if there is a later commit in the main branch than the latest one in a branch? This could help ensure versions we don't forget to update  `hubUtils` when a version is released?

This also makes me think we should add another bullet to the release instruction in the `schemas` repo to make sure to update `hubUtils` with the released version of the schema and make a release of `hubUtils` also.

Reply:
> Would it be possible to also always check the main branch and to also flag if there is a later commit in the main branch than the latest one in a branch? This could help ensure versions we don't forget to update `hubUtils` when a version is released?

I think this is reasonable.

> This also makes me think we should add another bullet to the release instruction in the `schemas` repo to make sure to update `hubUtils` with the released version of the schema and make a release of `hubUtils` also.


Definitely an issue to open on the schemas repo after this is done.

Reply:
done in [0e51294](https://github.com/hubverse-org/hubUtils/pull/194/commits/0e51294d1c98be4e0dc03461965dca83c482e357)